mock: act as proxy

pohly opened the pull request:
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

By running the mock driver in proxy mode with the csi.sock made
available via another *listening* endpoint it becomes possible to run
the actual driver code elsewhere, for example embedded inside an
e2e.test binary.

The second endpoint can be a TCP port that can be reached from outside
of a Kubernetes cluster, either via a service or port-forwarding.

The same can be done with "socat UNIX-LISTEN:/csi/csi.sock,fork
TCP-LISTEN:9000,reuseport", but the implementation inside the mock
driver has some advantages compared to that:
- the same image as for normal mock driver testing can be used
- the mock driver keeps the second listen socket open at all times,
  so a client connecting to it doesn't get "connection refused"
  errors, which happens while socat hasn't forked to accept
  such a connection

**Special notes for your reviewer**:

A corresponding change to Kubernetes is here: https://github.com/kubernetes-csi/node-driver-registrar/pull/124

**Does this PR introduce a user-facing change?**:
```release-note
mock driver: optionally proxy connections instead of handling them directly
```

msau42 commented:
/assign @jsafrane @tsmetana

pohly commented:
We discussed a bit whether this functionality is needed and if so, where it should be implemented: https://github.com/kubernetes/kubernetes/pull/97069#discussion_r566146552

jsafrane commented on internal/proxy/proxy_test.go:
```diff
@@ -0,0 +1,122 @@
+limitations under the License.
+*/
+
+// Package proxy makes it possible to forward a listening socket in
```
Nit: is it necessary to repeat the documentation in a test file?

pohly replied:
No, that wasn't intentional. Will remove it.

jsafrane commented on internal/proxy/proxy.go:
```diff
@@ -0,0 +1,148 @@
+				klog.V(5).Infof("proxy endpoint %s closed, shutting down", endpoint1)
+				return
+			}
+			go func() {
```
It took me a while to understand how this work. A separate goroutine for `s2` is started to start accepting on `s1` as soon as possible. OK, it looks working (a comment would be nice).

At the same time, the comment above says that kernel is accepting connections on listening sockets. Therefore, can the proxy call `accept(s2)` here, without a goroutine? Kernel will keep incoming connection on `s1`.

pohly replied:
The way it is now we collect connections from s1 as quickly as they come in and only rely on the kernel for s2. But it is indeed asymmetric and harder to understand. I'll try with a simpler approach.

pohly replied:
The unnecessary complex approach also had a race condition: conn1 might change between the two accepts.

pohly commented on internal/proxy/proxy.go:
```diff
@@ -0,0 +1,148 @@
+				klog.V(3).Infof("proxy established a new connection between %s and %s", endpoint1, endpoint2)
+				go copy(conn1, conn2, endpoint1, endpoint2)
+				go copy(conn2, conn1, endpoint2, endpoint1)
+			}()
```
This second `go copy` is where an unnecessary goroutine is created. I'll change that.

I also noticed a race condition: because `conn2` is bound to the variable outside the goroutine, it could change between the first accept and the second accept. It has to be a function parameter, and then a normal function makes more sense.

pohly commented:
@jsafrane: I've pushed the simplified proxy code, fixed the proxy_test.go comment and modified log levels slightly: when run at -v3, all we'll get from the container is the "proxy established a new connection between ..." message.

gRPC message logging will come from the embedded mock driver inside the e2e.test binary and the other sidecars (in particular, node-driver-registrar).

pohly commented:
I've verified that the e2e.test still works with this: tests are still running, but it worked fine so far.

jsafrane commented:
/lgtm
